### DISCUSSION TITLE
Adding Extendability into NetworkX

### DISCUSSION THREAD

**Discussion by gsemer** [networkx/networkx#4887] [Algorithms]
Hi all,
There is an algorithm that i came across during my diploma thesis and implementation was done by using networkx. I beleive that it might be of interest adding it into networkx.

Definition 
A graph G is k-extendable iff (i) G has a perfect matching and (ii) every matching of cardinality k extends to a perfect matching.

Extendability Problem
------------------------
Input: A graph G and a positive integer k.
Output: Is G k-extendable?
------------------------

In general case the above problem is co-NP-complete([2]).
If graph G is bipartite, then it can be decided in polynomial time([1]).

Let G = (V,E) be a simple, connected, undirected and bipartite graph with a perfect matching M and bipartition (U,V).
An edge is said to be saturated if it belongs to M and unsaturated otherwise.
We call residual graph of G, denoted by G_M, the graph obtained from G by directing the unsaturated edges from U to V and the saturated edges from V to U. 

Lemma([1])
Let M be a perfect matching of G. G is k-extendable iff its residual graph G_M is strongly connected and there are k vertex-disjoint directed paths between every vertex of U and every vertex of V.

Do you think networkx would benefit from it being added to the bipartite module?

References
------------
..[1] "A polynomial algorithm for the extendability problem in bipartite graphs", J. Lakhal, L. Litzler, Information Processing Letters, 1998.
..[2] "The matching extension problem in general graphs is co-NP-complete", Jan Hackfeld, Arie M. C. A. Koster, Springer Nature, 2018.
------------

**Comment by dschult** [comment]
Thank you for your offer of this interesting algorithm. The description you provide is focused, quite reasonably, on the results you created for your thesis. These results describe the complexity of the runtime in the limit of large graphs. 

Can you describe a little more about why someone would want to use this function? I guess that it identifies whether a graph is k-extendable. What does the user then do with this information? What is the bigger picture?

**Reply by gsemer** [reply]
Thank you for your immediate response. 
I read the message five to ten times before sending it and still forgot to mention the most important information. I am really sorry.
This algorithm accepts as input a graph G and returns its extendability.
The extendability of a graph G, denoted by ext(G), is defined as the maximum value of k such that G is k-extendable, i.e. 
ext(G) = max({k | G is k-extendable}).
Thus, we concentrate on the maximization version of the aforementioned problem. 
Obviously, there is a limitation on the choice of graphs as you correctly noticed. Since the runtime of the algorithm is O(n^3 * m^2), we are not able to use large graphs as inputs for this algorithm. ​

Hope it helps! Please, let me know if you  stillwant some further details.

References
------------
..[3] "On n-extendable graphs", M.D. Plummer, Discrete Math. 31 (1980) 201-210

**Reply by dschult** [reply]
Yes, this would be of interest to NetworkX. Maybe start by adding a module to the algorithms/bipartite directory along with a test module in the tests subdirectory there.  If it's appropriate we could move it to the matching.py module, but I think it will be better (and easier to review/merge etc) on its own.  You can open a pull request even if it isn't ready for release -- that way we can provide feedback along the way, and we won't merge until its ready.

Thanks!

**Reply by gsemer** [reply]
I just created a pull request for the extendability.py file. I continue with test_extendability.py. It seems that the code works correctly but i would like to use some new inputs.

**Reply by gsemer** [reply]
Thank you for your interest!

**Reply by gsemer** [reply]
Since it seems that the code works fine, you can merge it whenever you want. Unless, of course,  you would like to discuss something about the code.
I would like to ask something important.
Do you mind if i mention in my diploma that i implement the aforementioned algorithm using networkx tool?
I was thinking that it would be nice to add a section in my thesis with the name "Implementation in Python using NetworkX tool" and describe the procedure and parts of the code.
